### PR TITLE
feat: allow managing of personal access tokens

### PR DESCRIPTION
**Pull request by mediremi**
Closes https://jira.dhis2.org/browse/DHIS2-11649

This PR adds the ability to create, view and delete personal access tokens.

Potential security issues with this feature are highlighted in the app:
- Before generating a token, users are shown the complete list of authorities that the token will have access to
- We describe how IP address validation works (X-Forwarded-For) header and the potential for spoofing
- We explain that the Referer header should not be relied upon for access control
- We preselect 'GET' as the only allowed HTTP method

### CODE REVIEW

**Comment by varl** [comment]
Getting `400 Bad request` on debug/dev. Is this not available there ?

File: src/personalAccessTokens/generateTokenModal/ContextSelector.component.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+            >
+                {i18n.t('Use cases: Integrations and scripts', {
+                    nsSeparator: '-:-',
+                })}
+            </Context>
+            <Context
+                title={i18n.t('Browser context')}
+                onClick={() => setContext('BROWSER')}
+            >
+                {i18n.t('Use cases: Public portals', { nsSeparator: '-:-' })}
```
[the remark this answers is not in the document]
Comment (varl) [reply src/personalAccessTokens/generateTokenModal/ContextSelector.component.js:39]: I don't think it belongs in the `ContextSelector`. I _almost_ want it to pop up in a confirm/cancel dialog when I click on "Generate token" to make sure the user understands the risks.

[the remark this answers is not in the document]
Comment (mediremi) [reply src/personalAccessTokens/generateTokenModal/ContextSelector.component.js:39]: Would the following text work for the confirm dialog warning?

> Personal access tokens should only be used in a browser context for public access instances. For private instances, tokens should be treated like passwords and be kept private — requests should instead be routed through a proxy.

[the remark this answers is not in the document]
Comment (varl) [reply src/personalAccessTokens/generateTokenModal/ContextSelector.component.js:39]: Yeah, looks good to me. :+1:

**Comment by amcgee** [comment]
Looks good!  @cooper-joe might want to take a look

**Comment by amcgee** [comment]
One note - the relative times rendered for i.e. "Created X hours ago" do not take into account server-client timezone divergence - @ismay or @HendrikThePendric might have thoughts.

In this screenshot, I created the "2 hours ago" API key about 10 seconds before - my browser is in CET
Screen Shot 2021-08-30 at 23 26 00

**Review by varl — APPROVED** [review]

**Comment by cooper-joe** [comment]
>  @cooper-joe might want to take a look

Taking a look now, there's a few things I think could be adjusted. I'll write up some notes for my suggestions here.

**Comment by cooper-joe** [comment]
# Design feedback
Here's some suggested changes for this feature. I think there's also some potential improvements to be made to how the newly created tokens are displayed, but that would require more time to look into.

## Modal
[Mockup/prototype of my suggestions](https://www.figma.com/file/xNVdSYiQV9TUZKYVfqfFYS/Users?node-id=1%3A641) for the token creation modal, which incorporates the suggestions below.
- use checkboxes for the *allowed http methods* rather than switches (switches are only used when indicating something is on as in "running", like a job).
- move the *will have the following authorities* notice box under the other controls, otherwise it's not clear there are any options.
- use custom radio button controls for the type of token selection. allow toggling back and forth of token type without needing to cancel the modal and open a new one.
- adjust help texts and labels for clarity - see prototype/mockup for suggestions.

## List
- separate tokens into two lists by type (script / browser), with headers.
- use `ui` button for "copy to clipboard" functionality with `IconCopy` and text label *Copy to clipboard*.
- Add a title to the successfully created token, such as *Newly created token*.

**Comment by mediremi** [comment]
Thanks for the mockup @cooper-joe :+1: 

Regarding separating the list into server/browser contexts, unfortunately this information is not stored in the server but we can infer it from the presence of referrer or IP address restrictions. Since tokens can be created without such restrictions, how do you think we should handle tokens of ambiguous contexts? An 'Other' section/header?

**Comment by cooper-joe** [comment]
> Regarding separating the list into server/browser contexts, unfortunately this information is not stored in the server but we can infer it from the presence of referrer or IP address restrictions. Since tokens can be created without such restrictions, how do you think we should handle tokens of ambiguous contexts? An 'Other' section/header?

Without a reliable indicator, I think it's probably best not to separate the two lists. The "Other" section would be quite confusing, I think. Have removed that from the suggestions.

**Pull request merged**